tools/wptserve/wptserve/handlers.py

```diff
@@ -276,6 +277,29 @@ def get_range_data(self, f, byte_range):
 file_handler = FileHandler()  # type: ignore
 
 
+class HtmlScriptInjectorHandlerWrapper:
+    def __init__(self, inject="", wrap=None):
+        self.inject = inject
+        self.wrap = wrap
+
+    def __call__(self, request, response):
+        self.wrap(request, response)
+        # If the response content type isn't html, don't modify it.
+        if not isinstance(response.headers, ResponseHeaders) or response.headers.get("Content-Type")[0] != b"text/html":
+            return response
+
+        # Otherwise, inject the polyfill script after the document's doctype.
+        data = "".join(item.decode(response.encoding) for item in response.iter_content(read_file=True))
+        response.content = re.sub(r'^(<![^>]*>\n?)?',
```

Comment:
It's going to be hard to make any regexp based solution reliable, but this is going to fail to match if the `<` isn't literally the first byte in the file, and do something unexpected for leading comments e.g. `<!-- comment --><!DOCTYPE html>`.

Maybe this is overkill, but since we've vendored html5lib, one could "just" tokenize until one has either seen the doctype or a token, and then insert after/before that token i.e. something like (as a standalone script):

```py
#!/usr/bin/env python3

import sys

from html5lib import html5parser


def main(data, inject_data):
    token_types = html5parser.tokenTypes
    after_tokens = {token_types["Doctype"]}

    before_tokens = {token_types["StartTag"], token_types["EndTag"],
                     token_types["EmptyTag"], token_types["Characters"]}

    error_tokens = {token_types["ParseError"]}

    tokenizer = html5parser._tokenizer.HTMLTokenizer(data)
    stream = tokenizer.stream
    offset = 0
    error = False
    for item in tokenizer:
        if item["type"] in before_tokens:
            break
        elif item["type"] in error_tokens:
            error = True
            break

        offset = stream.chunkOffset

        if item["type"] in after_tokens:
            break
    else:
        error = True

    if not error and stream.prevNumCols or stream.prevNumLines:
        # We're outside the first chunk, so we don't know what to do
        error = True

    if error:
        return data

    return data[:offset] + inject_data + data[offset:]


if __name__ == "__main__":
    print(main(sys.argv[1], "<script>Insert polyfill here</script>"))
```

That will be slower, but should work in more cases.

Reply:
Done, I switched to just using before_tokens since inserting after the doctype also meant inserting before any following spaces. The spec recommends that the doctype is followed by a newline: https://www.w3.org/TR/2011/WD-html5-20110525/syntax.html#writing